osmnx/utils.py

```diff
@@ -573,6 +618,125 @@ def get_nearest_nodes(G, X, Y, method=None):
+    -------
+    closest_road_to_point : tuple
+        A geometry object representing the segment and the coordinates of the two
+        nodes that determine the road section, u and v, the OSM ids of the nodes.
+    """
+    start_time = time.time()
+
+    if method is None:
+        # calculate nearest edge one at a time for each point
+        ne = [get_nearest_edge(G, (x, y)) for x, y in zip(X, Y)]
+        ne = [(u, v) for _, u, v in ne]
+
+    elif method == 'kdtree':
+
+        # check if we were able to import scipy.spatial.cKDTree successfully
+        if not cKDTree:
+            raise ImportError('The scipy package must be installed to use this optional feature.')
+
+        # transform graph into DataFrame
+        edges = graph_to_gdfs(G, nodes=False, fill_edge_geometry=True)
+
+        # transform edges into evenly spaced points
+        edges['points'] = edges.apply(lambda x: redistribute_vertices(x.geometry, dist), axis=1)
+
+        # develop edges data for each created points
+        extended = edges['points'].apply([pd.Series]).stack().reset_index(level=1, drop=True).join(edges).reset_index()
+
+        # Prepare btree arrays
+        nbdata = np.array(list(zip(extended['Series'].apply(lambda x: x.x),
+                                   extended['Series'].apply(lambda x: x.y))))
+
+        # build a k-d tree for euclidean nearest node search
+        btree = cKDTree(data=nbdata, compact_nodes=True, balanced_tree=True)
+
+        # query the tree for nearest node to each point
+        points = np.array([X, Y]).T
+        dist, idx = btree.query(points, k=1)  # Returns ids of closest point
+        eidx = extended.loc[idx, 'index']
+        ne = edges.loc[eidx, ['u', 'v']]
+
+    else:
+        raise ValueError('You must pass a valid method name, or None.')
+
+    log('Found nearest edges to {:,} points in {:,.2f} seconds'.format(len(X), time.time() - start_time))
+
+    return np.array(ne)
+
+
+def redistribute_vertices(geom, dist):
+    """
+    Redistribute the vertices on a projected LineString or MultiLineString. The distance
+    argument is only approximate since the total distance of the linestring may not be
+    a multiple of the preferred distance. This function works on only [Multi]LineString
+    geometry types.
+
+    This code is adapted from an answer by Mike T from this original question:
+    https://stackoverflow.com/questions/34906124/interpolating-every-x-distance-along-multiline-in-shapely
+
+    Parameters
+    ----------
+    geom: LineString or MultiLineString
+        a Shapely geometry
+    dist : float
+        point spacing along edges. The smaller, the more points are created.
+        Units must
+
+    Returns
+    -------
+        list of Point geometries : list
+    """
+    if geom.geom_type == 'LineString':
+        num_vert = int(round(geom.length / dist))
+        if num_vert == 0:
+            num_vert = 1
+        return [geom.interpolate(float(n) / num_vert, normalized=True)
```

Comment:
I would suggest using `none` as the default method as this works best for unprojected graphs (which are themselves the default) and does not require scipy, which is an optional dependency for OSMnx. A two-dimensional kdtree works only approximately for spheroidal lat-long data, and there should be a docstring comment here recommending it only be used on projected graphs.

Another option here would be to use scikit-learn's balltree with the haversine distance metric to more faithfully measure distances along the earth's curved surface. This would however require (another) new optional dependency in scikit-learn.

Reply:
I pushed some changes to address your comments:
- Reverted to None method being the default one + better docstrings
- New: Added balltree method for unprojected graphs
- Adapted tests for new balltree method